include/logging/log_msg2.h

```diff
@@ -34,6 +34,22 @@ typedef uint64_t log_timestamp_t;
 typedef uint32_t log_timestamp_t;
 #endif
 
+/* Determine if __auto_type is supported. If not then runtime approach must always
+ * be used.
+ *
+ * Supported by:
+ * - GCC 4.9.0 https://gcc.gnu.org/gcc-4.9/changes.html
+ * - Clang 3.8
+ */
+#ifndef CONFIG_LOG2_ALWAYS_RUNTIME
+#if !defined(__cplusplus) && \
+	!(((__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__) >= 40900) || \
+	((__clang_major__ * 10000 + __clang_minor__ * 100 + __clang_patchlevel__) >= 30800))
```

Comment:
Maybe put this in a #define ?

#define AUTO_TYPE_IS_SUPPORTED

Reply:
are there any way we could detect if this is supported before build time ?

For example checking compiler version in CMake or testing the compiler (I'm not aware of any compiler flags related to this feature, so this will most likely not be possible).

Reply:
not sure about compile time, but i rephrased this define to fixed https://github.com/zephyrproject-rtos/zephyr/pull/42686#discussion_r806654572 same time.

Reply:
placed comment here: https://github.com/zephyrproject-rtos/zephyr/pull/42686#discussion_r811090728

comments to comments easily gets lost in history threads.